File: camerax_android/lib/src/core/image_capture_impl.dart
import 'dart:ui' as ui;

import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/common.dart';
import 'package:camerax_android/src/core.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class ImageCaptureOnImageCapturedCallbackImpl
    implements ImageCapture$OnImageCapturedCallback {
  final ImageCaptureOnImageCapturedCallbackProxyApi api;

  ImageCaptureOnImageCapturedCallbackImpl.internal(this.api);
}

final class ImageCaptureChannelImpl extends ImageCaptureChannel {
  @override
  ImageCapture$OnImageCapturedCallback createOnImageCapturedCallback({
    void Function()? onCaptureStarted,
    void Function(int progress)? onCaptureProcessProgressed,
    void Function(ui.Image bitmap)? onPostviewBitmapAvailable,
    void Function(ImageProxy image)? onCaptureSuccess,
    void Function(Object exception)? onError,
  }) {
    final api = ImageCaptureOnImageCapturedCallbackProxyApi(
      onCaptureStarted: onCaptureStarted == null
          ? null
          : (_) => onCaptureStarted(),
      onCaptureProcessProgressed: onCaptureProcessProgressed == null
          ? null
          : (_, e) => onCaptureProcessProgressed(e),
      onPostviewBitmapAvailable: onPostviewBitmapAvailable == null
          ? null
          : (_, e) async {
              final image = await e.impl();
              onPostviewBitmapAvailable(image);
            },
      onCaptureSuccess: onCaptureSuccess == null
          ? null
          : (_, e) => onCaptureSuccess(e.impl),
      onError: onError == null ? null : (_, e) => onError(e.impl),
    );
    return ImageCaptureOnImageCapturedCallbackImpl.internal(api);
  }
}

extension ImageCapture$CaptureModeX on ImageCapture$CaptureMode {
  ImageCaptureCaptureModeApi get api =>
      ImageCaptureCaptureModeApi.values[index];
}

extension ImageCaptureCaptureModeApiX on ImageCaptureCaptureModeApi {
  ImageCapture$CaptureMode get impl => ImageCapture$CaptureMode.values[index];
}

extension ImageCapture$FlashModeX on ImageCapture$FlashMode {
  ImageCaptureFlashModeApi get api => ImageCaptureFlashModeApi.values[index];
}

extension ImageCaptureFlashModeApiX on ImageCaptureFlashModeApi {
  ImageCapture$FlashMode get impl => ImageCapture$FlashMode.values[index];
}

extension ImageCapture$OnImageCapturedCallbackX
    on ImageCapture$OnImageCapturedCallback {
  ImageCaptureOnImageCapturedCallbackProxyApi get api {
    final impl = this;
    if (impl is! ImageCaptureOnImageCapturedCallbackImpl) throw TypeError();
    return impl.api;
  }
}

File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/core/ImageCaptureImpl.kt
package dev.zeekr.camerax_android.core

import android.graphics.Bitmap
import androidx.annotation.OptIn
import androidx.camera.core.ExperimentalZeroShutterLag
import androidx.camera.core.ImageCapture
import androidx.camera.core.ImageCaptureException
import androidx.camera.core.ImageProxy
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.ImageCaptureCaptureModeApi
import dev.zeekr.camerax_android.ImageCaptureFlashModeApi
import dev.zeekr.camerax_android.PigeonApiImageCaptureOnImageCapturedCallbackProxyApi
import dev.zeekr.camerax_android.common.api

class ImageCaptureImpl {
    class OnImageCapturedCallbackImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
        PigeonApiImageCaptureOnImageCapturedCallbackProxyApi(registrar) {
        override fun pigeon_defaultConstructor(): ImageCapture.OnImageCapturedCallback {
            return object : ImageCapture.OnImageCapturedCallback() {
                override fun onCaptureStarted() {
                    super.onCaptureStarted()
                    onCaptureStarted(this) {}
                }

                override fun onCaptureProcessProgressed(progress: Int) {
                    super.onCaptureProcessProgressed(progress)
                    onCaptureProcessProgressed(this, progress.toLong()) {}
                }

                override fun onPostviewBitmapAvailable(bitmap: Bitmap) {
                    super.onPostviewBitmapAvailable(bitmap)
                    onPostviewBitmapAvailable(this, bitmap) {}
                }

                override fun onCaptureSuccess(image: ImageProxy) {
                    super.onCaptureSuccess(image)
                    onCaptureSuccess(this, image) {}
                }

                override fun onError(exception: ImageCaptureException) {
                    super.onError(exception)
                    onError(this, exception.api) {}
                }
            }
        }
    }
}

val Int.imageCaptureCaptureModeApi: ImageCaptureCaptureModeApi
    @OptIn(ExperimentalZeroShutterLag::class) get() = when (this) {
        ImageCapture.CAPTURE_MODE_MAXIMIZE_QUALITY -> ImageCaptureCaptureModeApi.MINIMIZE_LATENCY
        ImageCapture.CAPTURE_MODE_MINIMIZE_LATENCY -> ImageCaptureCaptureModeApi.MINIMIZE_LATENCY
        ImageCapture.CAPTURE_MODE_ZERO_SHUTTER_LAG -> ImageCaptureCaptureModeApi.MINIMIZE_LATENCY
        else -> throw IllegalArgumentException()
    }

val Int.imageCaptureFlashModeApi: ImageCaptureFlashModeApi
    get() = when (this) {
        ImageCapture.FLASH_MODE_AUTO -> ImageCaptureFlashModeApi.AUTO
        ImageCapture.FLASH_MODE_ON -> ImageCaptureFlashModeApi.ON
        ImageCapture.FLASH_MODE_OFF -> ImageCaptureFlashModeApi.OFF
        ImageCapture.FLASH_MODE_SCREEN -> ImageCaptureFlashModeApi.SCREEN
        else -> throw IllegalArgumentException()
    }

val ImageCaptureCaptureModeApi.impl: Int
    @OptIn(ExperimentalZeroShutterLag::class) get() = when (this) {
        ImageCaptureCaptureModeApi.MAXIMIZE_QUALITY -> ImageCapture.CAPTURE_MODE_MAXIMIZE_QUALITY
        ImageCaptureCaptureModeApi.MINIMIZE_LATENCY -> ImageCapture.CAPTURE_MODE_MINIMIZE_LATENCY
        ImageCaptureCaptureModeApi.ZERO_SHUTTER_LAG -> ImageCapture.CAPTURE_MODE_ZERO_SHUTTER_LAG
    }

val ImageCaptureFlashModeApi.impl: Int
    get() = when (this) {
        ImageCaptureFlashModeApi.AUTO -> ImageCapture.FLASH_MODE_AUTO
        ImageCaptureFlashModeApi.ON -> ImageCapture.FLASH_MODE_ON
        ImageCaptureFlashModeApi.OFF -> ImageCapture.FLASH_MODE_OFF
        ImageCaptureFlashModeApi.SCREEN -> ImageCapture.FLASH_MODE_SCREEN
    }
